### ISSUE TITLE
Performance drop in get_frame_async

### ISSUE THREAD

**Issue by Ichunjo** [vapoursynth/vapoursynth#785]
```py
clip = core.lsmas.LWLibavSource(path)
# clip[:1000].set_output(0)

import time
start = time.time()
with open(os.devnull, 'wb') as devnull:
    clip[:1000].output(devnull, True)
print()
duration = time.time() - start
print("Duration: ", duration)
```
```
# APIv3
# VideoNode.output
# Duration:  2.153500556945801
# 
# vspipe -p test3.py .
# Output 1000 frames in 3.16 seconds (316.47 fps)



# APIv4
# VideoNode.output
# Duration:  8.244498491287231
# 
# vspipe -p test3.py .
# Output 1000 frames in 2.13 seconds (469.05 fps)
```

VideoNode.output is much slower in api v4 than v3 while the vspipe output is similar minor the performance boost.

**Comment by theChaosCoder** [comment]
lsmas is still api3 I think, try with ffms2 again.

**Comment by Ichunjo** [comment]
I tried with a blank clip and the results are the same.
vspipe durations are mostly the same while `clip.output` has a significant performance drop in APIv4

**Comment by Ichunjo** [comment]
Another test on clip.frames():
```py
clip = core.std.BlankClip(width=3840, height=2160, format=vs.YUV420P8, length=1000)

import time
start = time.time()
for f in clip.frames():
    f.copy()

print()
duration = time.time() - start
print("Duration: ", duration)

# V4 Duration:  10.208998918533325
# V3 Duration:  0.448500394821167
```

**Comment by myrsloik** [comment]
I don't think I changed anything there. Odd.

**Comment by Ichunjo** [comment]
Correct me if I'm wrong but I think these PRs changed the behaviour of the frames requesting?
https://github.com/vapoursynth/vapoursynth/pull/727
https://github.com/vapoursynth/vapoursynth/pull/735

**Comment by myrsloik** [comment]
Right, that probably triggers full frame copies. Let's summon @Kamekameha 
The easiest solution is to change the function to use raw pointer instead of fancy python wrappers. I guess?

**Comment by Kamekameha** [comment]
Frame requests?  I don't see how my changes would affect any of that.  The changes are present in both APIv3 and v4 branches too, so, if anything, I'd expect the same slowdown in both branches if truly the culprit.

What does change between branches, however, is the content of `VideoNode.output()`, which in turn now uses an also updated `VideoNode.frames()` method and ultimately `VideoNode.get_frame_async()`, and therefore I'm more inclined to believe the performance hit stems from there, as suggested by the current issue's title.

I suspect that https://github.com/vapoursynth/vapoursynth/blob/ec3407dc9ae3206041fb821d6b154718d54259a4/src/cython/vapoursynth.pyx#L656-L657 combined with https://github.com/vapoursynth/vapoursynth/blob/ec3407dc9ae3206041fb821d6b154718d54259a4/src/cython/vapoursynth.pyx#L644-L645 are introducing some noticeable overhead.

In short, I investigated myself and found that I did nothing wrong.

**Comment by Kamekameha** [comment]
Okay, can confirm.  I did away with `CallbackData` and used a simpler object, and skipped on pushing the environment.  On my shitty laptop I get 334.12 fps versus previously 28.15 fps.

Putting my disagreements with this whole Environment thingy aside, I believe we shouldn't really bother in preserving it for `VideoNode.get_frame_async()`'s futures.  The only reason I see this is done is so that callbacks added with `concurrent.futures.Future.add_done_callback()` are called with the same environment set at the time.

I see little benefit to this, the world might look entirely different for any TLS data the callback might use, and if no callback was added to begin with then all the bookkeeping was for nothing.  I consider the burden of keeping such or any data should be moved to the user, specially when they can provide any callable they like already.

**Comment by myrsloik** [comment]
Sigh, you python users. I suppose the performance regression in frames() is a real problem but WHY DO YOU KEEP USING OUTPUT?
I've tried to deprecate it for years now.

**Comment by Ichunjo** [comment]
`clip.output` is the easiest way to render the clip and thus write the clip inside a python script:
- By doing just `python script.py` you can launch the command with the x264/5/ffmpeg commands stored inside the script.

- Using `vspipe` is too restrictive and it doesn’t load the submodules of the current script.
Consider this sctructure:
   ```
   project/
   │
   ├── project_common/
   │   ├── __init__.py
   │   ├── config.py
   │   └── filters.py
   │
   ├── enco_script.py
   ```
   Inside `enco_script.py` we import `project_common` and we use `clip.output`
   ```py
   from project_common import *
   ...
   clip.output(...)
   ```
   You can  `python enco_script.py` and it will works.

   However:
   ```py
   from project_common import *
   ...
   clip.set_output()
   ```
   `vspipe enco_script.py - ...` will give you an error because it doesn’t see the `project_common` module.

Of course [this](https://github.com/Irrational-Encoding-Wizardry/lvsfunc/blob/master/lvsfunc/render.py#L64#L191) is possible (although it relies on `get_frame_async` which is currently slow) but it’s a bit slower than the code in cython.

**Comment by Ichunjo** [comment]
Any chance you could make a PR to fix it @Kamekameha?
Having `get_frame_async`, `frames` and `output` that slow are currently the things that prevent me from using APIv4 for real encoding because I’m using all the time `output` for all my automation system…

**Comment by Kamekameha** [comment]
Right. Shouldn't we have @stuxcrystal chime in about this, though?  It's his code after all.

But yeah, if we're onboard with my suggestion then sure, I can produce a PR.

> WHY DO YOU KEEP USING OUTPUT?

`output()` is fine*, fite me.  <sub><sub>*It just need some minor rework.</sub></sub>

**Comment by myrsloik** [comment]
He's been missing for almost a year now so don't expect chiming to happen

**Comment by cid-chan** [comment]
Hi. I was once stuxcrystal.

I have been asked to work on it by other people, linking to this issue. I''ve been out of business about vapoursynth for almost a year now, so I'll need some time to catch up.

P.S. Please refer to me as "she" now. :transgender_flag:
